src/applications/drydock/blueprint/DrydockWorkingCopyBlueprintImplementation.php

```diff
@@ -0,0 +1,632 @@
+        $src_ref,
+        $src_ref);
+
+      $error = DrydockCommandError::newFromCommandException($ex)
+        ->setPhase(self::PHASE_MERGEFETCH)
+        ->setDisplayCommand($display_command);
+
+      $lease->setAttribute('workingcopy.vcs.error', $error->toDictionary());
+
+      throw $ex;
+    }
+
+    // TM CHANGES BEGIN
+    // We try to rebase the diff off the branch we are trying land on.
+    // this means we can drop any commits which have already been landed.
+    $tmp_branch_name = sprintf('tmp-branch-%d', rand());
+    try {
+      $interface->execx(
+        'git checkout -b %s %s',
+        $tmp_branch_name,
+        $src_ref);
+    } catch (CommandException $ex) {
+      $display_command = cprintf(
+        'git checkout -b %s %s',
+        $tmp_branch_name,
+        $src_ref);
+      $error = DrydockCommandError::newFromCommandException($ex)
+        ->setPhase(self::PHASE_MERGEFETCH)
+        ->setDisplayCommand($display_command);
+      $lease->setAttribute('workingcopy.vcs.error', $error->toDictionary());
+
+      throw $ex;
+    }
+
+    try {
+      $interface->execx('git rebase -');
+    } catch (CommandException $ex) {
+      $interface->execx(
+        'git branch -D %s',
+        $tmp_branch_name);
+
+      $error = DrydockCommandError::newFromCommandException($ex)
+        ->setPhase(self::PHASE_MERGEREBASE)
+        ->setDisplayCommand('git rebase -');
+      $lease->setAttribute('workingcopy.vcs.error', $error->toDictionary());
+
+      throw $ex;
+    }
+
+    try {
+      $interface->execx('git checkout -');
+    } catch (CommandException $ex) {
+      $interface->execx(
+        'git branch -D %s',
+        $tmp_branch_name);
+      
+      $error = DrydockCommandError::newFromCommandException($ex)
+        ->setPhase(self::PHASE_MERGEREBASE)
+        ->setDisplayCommand('git checkout -');
+      $lease->setAttribute('workingcopy.vcs.error', $error->toDictionary());
+
+      throw $ex;
+    }
+
+    // NOTE: This can never actually generate a commit because we pass
+    // "--squash", but git sometimes runs code to check that a username and
+    // email are configured anyway.
+    $real_command = csprintf(
+      'git -c user.name=%s -c user.email=%s merge --no-stat --squash -- %R',
+      'drydock',
+      'drydock@phabricator',
+      $tmp_branch_name);
+
+    try {
+      $interface->execx('%C', $real_command);
+    } catch (CommandException $ex) {
+      $interface->execx(
+        'git branch -D %s',
+        $tmp_branch_name);
+      
+      $display_command = csprintf(
+        'git merge --squash %R',
+        $src_ref);
+
+      $error = DrydockCommandError::newFromCommandException($ex)
+        ->setPhase(self::PHASE_SQUASHMERGE)
+        ->setDisplayCommand($display_command);
+
+      $lease->setAttribute('workingcopy.vcs.error', $error->toDictionary());
+      throw $ex;
+    }
+
+    $interface->execx(
+      'git branch -D %s',
+      $tmp_branch_name);
+
+    // TM CHANGES END
```

Comment:
its a little difficult to see but this is the actual diff 
```
+    // TM CHANGES BEGIN
+    // We try to rebase the diff off the branch we are trying land on.
+    // this means we can drop any commits which have already been landed.
+    $tmp_branch_name = sprintf('tmp-branch-%d', rand());
+    try {
+      $interface->execx(
+        'git checkout -b %s %s',
+        $tmp_branch_name,
+        $src_ref);
+    } catch (CommandException $ex) {
+      $display_command = cprintf(
+        'git checkout -b %s %s',
+        $tmp_branch_name,
+        $src_ref);
+      $error = DrydockCommandError::newFromCommandException($ex)
+        ->setPhase(self::PHASE_MERGEFETCH)
+        ->setDisplayCommand($display_command);
+      $lease->setAttribute('workingcopy.vcs.error', $error->toDictionary());
+
+      throw $ex;
+    }
+
+    try {
+      $interface->execx('git rebase -');
+    } catch (CommandException $ex) {
+      $interface->execx(
+        'git branch -D %s',
+        $tmp_branch_name);
+
+      $error = DrydockCommandError::newFromCommandException($ex)
+        ->setPhase(self::PHASE_MERGEREBASE)
+        ->setDisplayCommand('git rebase -');
+      $lease->setAttribute('workingcopy.vcs.error', $error->toDictionary());
+
+      throw $ex;
+    }
+
+    try {
+      $interface->execx('git checkout -');
+    } catch (CommandException $ex) {
+      $interface->execx(
+        'git branch -D %s',
+        $tmp_branch_name);
+      
+      $error = DrydockCommandError::newFromCommandException($ex)
+        ->setPhase(self::PHASE_MERGEREBASE)
+        ->setDisplayCommand('git checkout -');
+      $lease->setAttribute('workingcopy.vcs.error', $error->toDictionary());
+
+      throw $ex;
+    }
 
     // NOTE: This can never actually generate a commit because we pass
     // "--squash", but git sometimes runs code to check that a username and
@@ -525,11 +579,15 @@ final class DrydockWorkingCopyBlueprintImplementation
       'git -c user.name=%s -c user.email=%s merge --no-stat --squash -- %R',
       'drydock',
       'drydock@phabricator',
-      $src_ref);
+      $tmp_branch_name);
 
     try {
       $interface->execx('%C', $real_command);
     } catch (CommandException $ex) {
+      $interface->execx(
+        'git branch -D %s',
+        $tmp_branch_name);
+      
       $display_command = csprintf(
         'git merge --squash %R',
         $src_ref);
@@ -541,6 +599,12 @@ final class DrydockWorkingCopyBlueprintImplementation
       $lease->setAttribute('workingcopy.vcs.error', $error->toDictionary());
       throw $ex;
     }
+
+    $interface->execx(
+      'git branch -D %s',
+      $tmp_branch_name);
+
+    // TM CHANGES END
```